plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/columns-control.tsx

```diff
@@ -2,89 +2,164 @@
  */
-import { DisplayLayoutControlProps } from '../../types';
-import { getDefaultDisplayLayout } from '../../constants';
+import {
+	LayoutOptions,
+	ProductCollectionLayoutGrid,
+	TemplateLayoutControlProps,
+} from '../../types';
+import {
+	DEFAULT_LAYOUT_GRID_COLUMNS,
+	DEFAULT_LAYOUT_GRID_MINIMUM_COLUMN_WIDTH,
+} from '../../constants';
 
+const autoColumnsLabel = __( 'Column Mode', 'woocommerce' );
+const autoModeLabel = __( 'Auto', 'woocommerce' );
+const manualModeLabel = __( 'Manual', 'woocommerce' );
+const minimumColumnWidthLabel = __( 'Minimum Column Width', 'woocommerce' );
 const columnsLabel = __( 'Columns', 'woocommerce' );
-const toggleLabel = __( 'Responsive', 'woocommerce' );
-const toggleHelp = __(
-	'Automatically adjust the number of columns to better fit smaller screens.',
-	'woocommerce'
-);
 
-const ColumnsControl = ( props: DisplayLayoutControlProps ) => {
-	const { type, columns, shrinkColumns } = props.displayLayout;
-	const showColumnsControl = type === 'flex';
+const ColumnsControl = ( props: TemplateLayoutControlProps ) => {
+	const templateLayout = props.templateLayout as ProductCollectionLayoutGrid;
 
-	const defaultLayout = getDefaultDisplayLayout();
+	// Rather than serializing whether or not they're using
+	// automatic columns we can just check the current
+	// layout. When they're using a minimum width we
+	// are in auto mode, and if nothing is set, we
+	// will just default to auto mode.
+	const [ autoColumns, setAutoColumns ] = useState(
+		templateLayout.minimumColumnWidth !== undefined ||
+			templateLayout.columnCount === undefined
+	);
 
-	const onShrinkColumnsToggleChange = ( value: boolean ) => {
+	const setTemplateLayout = (
+		newTemplateLayout: Partial< ProductCollectionLayoutGrid >
+	) => {
 		props.setAttributes( {
-			displayLayout: {
-				...props.displayLayout,
-				shrinkColumns: value,
+			templateLayout: {
+				type: LayoutOptions.GRID,
+				...newTemplateLayout,
 			},
 		} );
 	};
 
-	const onPanelDeselect = () => {
+	// Keep track of the last layout values so that we can switch back and forth
+	// without needing to serialize the previous value to the block.
+	const [ lastLayout, setLastLayout ] = useState(
+		autoColumns
+			? DEFAULT_LAYOUT_GRID_COLUMNS
+			: DEFAULT_LAYOUT_GRID_MINIMUM_COLUMN_WIDTH
+	);
+
+	const onAutoColumnsChange = ( value: boolean ) => {
+		// When we switch back and forth we should restore the last layout value
+		// to the template layout and store the opposite value in case we
+		// want to switch back to the other and not lose the value.
+		if ( value ) {
+			setTemplateLayout( {
+				minimumColumnWidth: lastLayout as string,
+			} );
+			setLastLayout( templateLayout.columnCount as number );
+		} else {
+			setTemplateLayout( {
+				columnCount: lastLayout as number,
+			} );
+			setLastLayout( templateLayout.minimumColumnWidth as string );
+		}
+
+		setAutoColumns( value );
+	};
+
+	const onColumnCountChange = ( value: number ) =>
```

Comment:
I think the flickering Manish raised in his PR comes from `setState` + `setAttributes` and multiple re-renders. If I understand correctly, multiple `useState` in this file are used only to preserve previous values of columns or min width. In that case I suggest using `useRef`.

From React docs: [How to get the previous props or state?](https://legacy.reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state):

> We have previously suggested a custom Hook called usePrevious to hold the previous value. However, we’ve found that most use cases fall into the two patterns described above. If your use case is different, you can [hold a value in a ref](https://legacy.reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables) and manually update it when needed. Avoid reading and updating refs during rendering because this makes your component’s behavior difficult to predict and understand.

I assume if we switch from state to ref, the flickering would be gone.

Reply:
Awesome, thank you for this! I made the change and the flickering went away.

Reply:
Actually, I take that back! I misunderstood what flickering we were talking about!